### PR TITLE
Basic auth script

### PR DESCRIPTION
**Pull request by albertodonato**
Add a script to wrap htpasswd to maintain users for http basic-auth.

The script will be called via juju run --application archive-auth-mirror to update the database on all units of the service.

### CODE REVIEW

**Comment by albertodonato** [comment]
[Add a script to manage users who have access to the archive](https://trello.com/c/uKzxlMbV/225-add-a-script-to-manage-users-who-have-access-to-the-archive)

**Comment by albertodonato** [comment]
[Add a script to manage users who have access to the archive](https://trello.com/c/uKzxlMbV/225-add-a-script-to-manage-users-who-have-access-to-the-archive)

File: README.md [hunk 1]
```diff
@@ -2,3 +2,20 @@
...
+
+## Managing basic authentication
+
+Credentials for basic authentication can be created with:
+
+```bash
+$ juju run --application archive-auth-mirror '/srv/archive-auth-mirror/bin/manage-user add <user> <pass>'
+```
+
+If the user is already present, its password will be updated.
```
Comment (sparkiegeek) [inline README.md]: s/its/their/

File: resources/manage-user [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+def get_parser():
+    parser = argparse.ArgumentParser(description=__doc__)
+    action = subparsers = parser.add_subparsers(
+        dest='action', help='action to perform')
+    action.required = True
+    add_action = subparsers.add_parser(
+        'add', help='add  a user. If the user exists, the password is updated')
+    add_action.add_argument('user', help='the username to add')
+    add_action.add_argument('password', help='the password for the user')
+    subparsers.add_parser('list', help='list users')
```
Comment (sparkiegeek) [inline resources/manage-user]: i think a bit of whitespace here would help legibility

File: unit_tests/test_manage_user.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+from pathlib import Path
+
+from fixtures import TestWithFixtures, TempDir
+
+from importlib.machinery import SourceFileLoader
+
+
+# manually import the script since it's not in a module
+manage_user = SourceFileLoader(
+    'manage_user', "resources/manage-user").load_module()
```
Comment (sparkiegeek) [inline unit_tests/test_manage_user.py]: hmm, why not keep it as a Python module and install it with the foo-bar name in the `shutil.copy`?

File: resources/manage-user [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    remove_action = subparsers.add_parser('remove', help='remove a user')
+    remove_action.add_argument('user', help='the user to remove')
+    return parser
+
+
+def add_user(auth_file, user, passwd):
+    '''Add a user.'''
+    auth_file.touch()  # ensure the file exists
+    subprocess.check_call(
+        ['htpasswd', '-b', str(auth_file), user, passwd],
```
Comment (sparkiegeek) [inline resources/manage-user]: Why don't we give password into stdin and use `-i`? That's what's recommended in htpasswd(1)

Comment (sparkiegeek) [inline resources/manage-user]: Hmm, let's not use the default password encryption of MD5 (!), how about `-B` for bcrypt?

File: resources/manage-user [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    add_action.add_argument('password', help='the password for the user')
+    subparsers.add_parser('list', help='list users')
+    remove_action = subparsers.add_parser('remove', help='remove a user')
+    remove_action.add_argument('user', help='the user to remove')
+    return parser
+
+
+def add_user(auth_file, user, passwd):
+    '''Add a user.'''
+    auth_file.touch()  # ensure the file exists
```
Comment (sparkiegeek) [inline resources/manage-user]: Instead of `touch()` here, a `-c` to `htpasswd` should do the trick?

File: resources/manage-user [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    subprocess.check_call(
+        ['htpasswd', '-b', str(auth_file), user, passwd],
+        stderr=subprocess.DEVNULL)
+
+
+def list_users(auth_file):
+    '''List existing users.'''
+    if not auth_file.exists():
+        return []
+    with open(str(auth_file)) as fh:
```
Comment (sparkiegeek) [inline resources/manage-user]: `with auth_file.open() as fh:`

File: README.md (hunk 1 above)
Comment (albertodonato) [inline README.md]: fixed

File: resources/manage-user (hunk 5 above)
Comment (albertodonato) [inline resources/manage-user]: no, it fails if the file is already there.

File: resources/manage-user (hunk 4 above)
Comment (albertodonato) [inline resources/manage-user]: it seems easier this way?

Comment (albertodonato) [inline resources/manage-user]: done

File: unit_tests/test_manage_user.py (hunk 3 above)
Comment (albertodonato) [inline unit_tests/test_manage_user.py]: I prefer to have all resources in one place, it's easier to track what gets copied.

File: resources/manage-user (hunk 4 above)
Comment (albertodonato) [inline resources/manage-user]: Anyway, done

File: resources/manage-user (hunk 6 above)
Comment (albertodonato) [inline resources/manage-user]: Done

File: README.md (hunk 1 above)
Comment (sparkiegeek) [inline README.md]: You did? I don't see it

File: resources/manage-user (hunk 5 above)
Comment (sparkiegeek) [inline resources/manage-user]: Ahh, fair enough - thanks

File: unit_tests/test_manage_user.py (hunk 3 above)
Comment (sparkiegeek) [inline unit_tests/test_manage_user.py]: OK

File: resources/manage-user [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    return parser
+
+
+def add_user(auth_file, user, passwd):
+    '''Add a user.'''
+    auth_file.touch()  # ensure the file exists
+    process = subprocess.Popen(
+        ['htpasswd', '-Bi', str(auth_file), user],
+        stdin=subprocess.PIPE, stderr=subprocess.PIPE)
+    _, err = process.communicate(passwd.encode('utf8'))
```
Comment (sparkiegeek) [inline resources/manage-user]: `subprocess.run(..., input=passwd.encode('utf-8'), check=True)` ?

File: resources/manage-user [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    subparsers.add_parser('list', help='list users')
+
+    remove_action = subparsers.add_parser('remove', help='remove a user')
+    remove_action.add_argument('user', help='the user to remove')
+
+    return parser
+
+
+def add_user(auth_file, user, passwd):
+    '''Add a user.'''
```
Comment (sparkiegeek) [inline resources/manage-user]: `s/'''/"""/`

File: resources/manage-user [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+def list_users(auth_file):
+    '''List existing users.'''
+    if not auth_file.exists():
+        return []
+    with auth_file.open() as fh:
+        return [line.split(':')[0] for line in fh]
+
+
+def remove_user(auth_file, user):
+    subprocess.check_call(
```
Comment (sparkiegeek) [inline resources/manage-user]: docstring please, with `"""`

File: resources/manage-user [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    process = subprocess.Popen(
+        ['htpasswd', '-Bi', str(auth_file), user],
+        stdin=subprocess.PIPE, stderr=subprocess.PIPE)
+    _, err = process.communicate(passwd.encode('utf8'))
+    if process.wait():
+        raise CommandFailed(err.decode('utf8'))
+
+
+def list_users(auth_file):
+    '''List existing users.'''
```
Comment (sparkiegeek) [inline resources/manage-user]: `s/'''/"""/`

File: unit_tests/test_manage_user.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+class ManageUserTest(TestWithFixtures):
+
+    def setUp(self):
+        super().setUp()
+        tempdir = self.useFixture(TempDir())
+        self.auth_file = Path(tempdir.path) / 'auth-file'
+
+    def test_add_user(self):
+        '''add_user adds the user with the specified password.'''
```
Comment (sparkiegeek) [inline unit_tests/test_manage_user.py]: docstring double quotes throughout

File: resources/manage-user (hunk 9 above)
Comment (albertodonato) [inline resources/manage-user]: fixed

**Pull request merged**